tests/gtest/tcp/tcp_sockopt.cc

```diff
@@ -0,0 +1,324 @@
+        return false;
+      }
+      m_pipe[ReadSide] = -1;
+    }
+
+    return write(m_pipe[WriteSide], "X", 1) == 1;
+  }
+};
+
+using sockopt_parameters = std::tuple<int, int, int, int>;
+using tcp_sockopt_positive = testing::TestWithParam<sockopt_parameters>;
+TEST_P(tcp_sockopt_positive, set_and_get_value) {
+  int socket_domain, level, optname, value;
+  std::tie(socket_domain, level, optname, value) = GetParam();
+
+  auto fd = reusable_cleanable_test_socket(socket_domain, static_cast<int>(SOCK_STREAM), 0);
+  EXPECT_GE(fd, 0) << "socket syscall failed";
+  auto result = setsockopt(fd, level, optname, &value, sizeof(value));
+  EXPECT_EQ(result, 0) << "setsockopt failed to set the value";
+
+  int actual_value = -1;
+  socklen_t actual_enable_len = sizeof(actual_value);
+  result = getsockopt(fd, level, optname, &actual_value, &actual_enable_len);
+  EXPECT_EQ(result, 0) << "getsockopt failed to get the value";
+  EXPECT_EQ(actual_enable_len, sizeof(actual_value))
+      << "Got unexpected size of agument";
+  ASSERT_EQ(actual_value, value);
+}
+
+/* The valid ranges are dictated by the Linux Kernel and not the TCP RFC 9293 */
+INSTANTIATE_TEST_CASE_P(
+    keep_alive, tcp_sockopt_positive,
+    testing::Values(std::make_tuple(AF_INET, SOL_SOCKET, SO_KEEPALIVE, 1),
+                    std::make_tuple(AF_INET, SOL_SOCKET, SO_KEEPALIVE, 0),
+                    std::make_tuple(AF_INET, IPPROTO_TCP, TCP_KEEPIDLE, 1),
+                    std::make_tuple(AF_INET, IPPROTO_TCP, TCP_KEEPIDLE,
+                                    std::numeric_limits<int16_t>::max()),
+                    std::make_tuple(AF_INET, IPPROTO_TCP, TCP_KEEPINTVL, 1),
+                    std::make_tuple(AF_INET, IPPROTO_TCP, TCP_KEEPINTVL,
+                                    std::numeric_limits<int16_t>::max()),
+                    std::make_tuple(AF_INET, IPPROTO_TCP, TCP_KEEPCNT, 1),
+                    std::make_tuple(AF_INET, IPPROTO_TCP, TCP_KEEPCNT,
+                                    std::numeric_limits<int8_t>::max()),
+                    std::make_tuple(AF_INET6, SOL_SOCKET, SO_KEEPALIVE, 1),
+                    std::make_tuple(AF_INET6, SOL_SOCKET, SO_KEEPALIVE, 0),
+                    std::make_tuple(AF_INET6, IPPROTO_TCP, TCP_KEEPIDLE, 1),
+                    std::make_tuple(AF_INET6, IPPROTO_TCP, TCP_KEEPIDLE,
+                                    std::numeric_limits<int16_t>::max()),
+                    std::make_tuple(AF_INET6, IPPROTO_TCP, TCP_KEEPINTVL, 1),
+                    std::make_tuple(AF_INET6, IPPROTO_TCP, TCP_KEEPINTVL,
+                                    std::numeric_limits<int16_t>::max()),
+                    std::make_tuple(AF_INET6, IPPROTO_TCP, TCP_KEEPCNT, 1),
+                    std::make_tuple(AF_INET6, IPPROTO_TCP, TCP_KEEPCNT,
+                                    std::numeric_limits<int8_t>::max())));
+
+using tcp_setsockopt_negative = testing::TestWithParam<sockopt_parameters>;
```

Comment:
Please consider adding test description as in referenced https://github.com/Mellanox/libvma/blob/master/tests/gtest/tcp/tcp_connect.cc#L138